src/internet_identity/src/archive.rs

```diff
@@ -0,0 +1,179 @@
+            let archive_canister_id = data.archive_canister;
+            state::persistent_state_mut(|persistent_state| {
+                // safe archive info permanently
+                persistent_state.archive_info.state = ArchiveState::Created(data)
+            });
+            match install_archive(archive_canister_id, wasm).await {
+                Ok(()) => DeployArchiveResult::Success(archive_canister_id),
+                Err(err) => UpgradeFailed(err),
+            }
+        }
+    }
+}
+
+async fn create_archive() -> CallResult<ArchiveData> {
+    let (result,) = create_canister(CreateCanisterArgument { settings: None }).await?;
+    Ok(ArchiveData {
+        sequence_number: 0,
+        archive_canister: result.canister_id,
+    })
+}
+
+async fn install_archive(archive_canister: Principal, wasm_module: Vec<u8>) -> Result<(), String> {
+    let expected_hash = verify_wasm_hash(&wasm_module)?;
+
+    let (archive_status,) = canister_status(CanisterIdRecord {
+        canister_id: archive_canister,
+    })
+    .await
+    .map_err(|err| format!("failed to retrieve archive status: {:?}", err))?;
+
+    let module_hash = archive_status.module_hash;
+    if module_hash
+        .clone()
+        .map(|hash| hash == expected_hash.to_vec())
+        .unwrap_or(false)
+    {
+        // Don't do anything further if the archive canister has the given module already installed
+        return Ok(());
+    }
+
+    let mode = match module_hash {
+        None => Install,
+        Some(_) => Upgrade,
+    };
+
+    let settings = ArchiveInit {
+        ii_canister: id(),
+        max_entries_per_call: 1000,
+    };
+    let encoded_arg = candid::encode_one(settings)
+        .map_err(|err| format!("failed to encode archive install argument: {:?}", err))?;
+
+    install_code(InstallCodeArgument {
+        mode,
+        canister_id: archive_canister,
+        wasm_module,
+        arg: encoded_arg,
+    })
+    .await
+    .map_err(|err| format!("failed to install archive canister: {:?}", err))
+}
+
+pub fn archive_operation(anchor: UserNumber, caller: Principal, operation: Operation) {
+    let archive_data = state::archive_data(); // traps if archive is not available
+    let timestamp = time();
+    let entry = Entry {
+        anchor,
+        operation,
+        timestamp,
+        caller,
+        sequence_number: archive_data.sequence_number,
+    };
+    let encoded_entry = candid::encode_one(entry).expect("failed to encode archive entry");
+
+    // Notify can still trap if the message cannot be enqueued rolling back the anchor operation.
+    // Therefore we only increment the sequence number after notifying successfully.
+    let () = notify(
+        archive_data.archive_canister,
+        "write_entry",
+        (anchor, timestamp, encoded_entry),
+    )
+    .expect("failed to send archive entry notification");
+    state::increment_archive_seq_nr();
+}
+
+fn verify_wasm_hash(wasm_module: &Vec<u8>) -> Result<[u8; 32], String> {
```

Comment:
out of curiosity, did you check the computation cost of this function and if spamming `deploy_archive` would have a noticeable impact on II (DoS)?

Reply:
So I did some quick profiling and indeed hashing takes quite a few cycles:
* Hashing the actual archive.wasm: 72'438'283 cycles
* Submit bogus payload close to message size limit: 243'362'366 cycles

For comparison:
* `lookup`: 1'824'387 cycles
* `prepare_delegation`: 2'198'374 cycles

So roughly, one bogus `deploy_archive` call is about 100 times more expensive than an authentication process. We could limit payload size to not check the hash if it is outside the expected range which would make it only 35 times more expensive.

Reply:
I've talked to @roman-kashitsyn about this: I'll first check the installed `wasm` module and if it is already matching the configured one, I'll reject the payload right away, as there is nothing to do anyway. This will make it super cheap, once the archive has been deployed.

Reply:
Ok, I have now refactored the code as outlined above and the results are as follows:
* After deploying the archive and having cached the archive status a call to `deploy_archive` uses now `8'747'312` cycles of which about >99%(!) can be attributed to just decoding the incoming message.

We could further optimize the cycles usage by not decoding the argument, but I would only do this if it indeed turns out to be a problem since it requires more low level argument handling.